File: policyguard/engine.py
"""The engine that turns log lines into verdicts.

This is where the three interesting data structures meet:

* a **hash table**, the automaton's transition dictionary, doing the matching
* a **queue**, the sliding window that decides when a threshold rule fires
* a **graph**, the automaton itself, whose path is reported with each verdict

It is also where precedence lives. Several rules can match one line, so one
of them has to win. The rule is simple and is stated in the README, the CLI
help, and here: the highest priority match wins, and rules of equal priority
break the tie by their order in the pack.

The important consequence is that an explicit allow outranks an alert by
default. That is the behaviour an operator expects from an allow list, and
it is also the sharpest edge in the design, so it is written down in all
three places rather than left to be discovered.
"""

from __future__ import annotations

from collections import deque
from typing import Final, Iterable, Iterator

from policyguard.alert import Alert, Verdict
from policyguard.automaton import Automaton
from policyguard.event import Event, events_from_lines
from policyguard.rule import Decision, Rule, RulePack

#: How many recent events a threshold rule keeps per source when its rule
#: does not set a time window. Bounding this is what stops a long log from
#: growing the engine's memory without limit.
DEFAULT_WINDOW_EVENTS: Final[int] = 50


class _SlidingWindow:
    """Recent hits for one threshold rule, grouped by source.

    The queue is the point. Every time the rule's pattern is seen the event's
    position joins the back; anything that has fallen outside the window
    leaves the front. The rule fires when the queue is as long as the
    threshold. Both ends are touched in constant time, which is exactly what
    a deque is for and exactly what a list would do badly.
    """

    __slots__ = ('_threshold', '_span', '_hits')

    def __init__(self, threshold: int, span: int) -> None:
        """Create a window.

        Parameters
        ----------
        threshold : int
            How many hits are needed before the rule fires.
        span : int
            How many events wide the window is.
        """
        self._threshold = threshold
        self._span = max(span, threshold)
        self._hits: dict[str, deque[int]] = {}

    def record(self, source: str, position: int) -> int:
        """Record a hit and report how many are currently in the window.

        Parameters
        ----------
        source : str
            The address the event came from. Counting per source is what
            keeps one noisy host from tripping a rule for everyone else.
        position : int
            The event's position in the stream, counting from one.

        Returns
        -------
        int
            How many hits from this source are inside the window now.
        """
        queue = self._hits.setdefault(source, deque())
        queue.append(position)

        # Drop anything that has slid out of the window. Cheap, because only
        # the front can ever be stale.
        while queue and position - queue[0] >= self._span:
            queue.popleft()

        return len(queue)

    def is_tripped(self, count: int) -> bool:
        """Report whether a hit count is enough to fire the rule.

        Parameters
        ----------
        count : int
            The number returned by :meth:`record`.

        Returns
        -------
        bool
            Whether the rule should fire.
        """
        return count >= self._threshold

    def __str__(self) -> str:
        """Return a one-line summary for debugging output."""
        tracked = sum(len(queue) for queue in self._hits.values())
        return (
            f'window threshold={self._threshold} span={self._span} '
            f'sources={len(self._hits)} tracked={tracked}'
        )

    def __repr__(self) -> str:
        """Return an unambiguous form for debugging."""
        return f'_SlidingWindow(threshold={self._threshold}, span={self._span})'


class PolicyEngine:
    """Classifies log lines against a rule pack.

    Build one per pack. The automaton is compiled once in the constructor,
    so classifying a second file costs nothing extra.
    """

    __slots__ = ('_pack', '_automaton', '_windows', '_alerts', '_counts', '_order')

    def __init__(self, pack: RulePack) -> None:
        """Compile a pack into a ready engine.

        Parameters
        ----------
        pack : RulePack
            The rules to enforce.
        """
        self._pack = pack
        self._automaton = Automaton().compile(pack.rules)
        self._windows: dict[str, _SlidingWindow] = {
            rule.rule_id: _SlidingWindow(
                threshold=rule.threshold,
                span=rule.window_seconds or DEFAULT_WINDOW_EVENTS,
            )
            for rule in pack.rules
            if rule.is_threshold_rule
        }
        # Where each rule sits in the pack. Two rules of equal priority break
        # their tie by this, which is what the README and the CLI help both
        # promise. Without it the winner would be whichever rule happened to
        # finish earliest in the line, which is not something a rule author
        # can see or reason about.
        self._order: dict[str, int] = {
            rule.rule_id: index for index, rule in enumerate(pack.rules)
        }
        self._alerts: list[Alert] = []
        self._counts: dict[Decision, int] = {decision: 0 for decision in Decision}

    @property
    def pack(self) -> RulePack:
        """RulePack: The rules this engine enforces."""
        return self._pack

    @property
    def automaton(self) -> Automaton:
        """Automaton: The compiled DFA, for inspection and DOT export."""
        return self._automaton

    @property
    def alerts(self) -> tuple[Alert, ...]:
        """tuple of Alert: Every alert raised since the last reset."""
        return tuple(self._alerts)

    @property
    def counts(self) -> dict[Decision, int]:
        """dict: How many lines fell into each decision since the last reset."""
        return dict(self._counts)

    def reset(self) -> None:
        """Forget every alert, count, and sliding window.

        Call this between files. Without it a threshold rule would carry
        yesterday's failed logins into today's report.
        """
        self._alerts.clear()
        self._counts = {decision: 0 for decision in Decision}
        for rule in self._pack.rules:
            if rule.is_threshold_rule:
                self._windows[rule.rule_id] = _SlidingWindow(
                    threshold=rule.threshold,
                    span=rule.window_seconds or DEFAULT_WINDOW_EVENTS,
                )

    def classify(self, event: Event, position: int = 0) -> Verdict:
        """Classify one event.

        Parameters
        ----------
        event : Event
            The line to classify.
        position : int, optional
            The event's position in the stream, used by the sliding windows.
            Defaults to the event's line number.

        Returns
        -------
        Verdict
            What the line was called, and why.
        """
        index = position or event.line_number
        winner: Rule | None = None
        winning_path: tuple[int, ...] = ()
        winning_hits = 1

        for match in self._automaton.match(event.tokens):
            rule = match.rule
            hits = 1

            if rule.is_threshold_rule:
                hits = self._windows[rule.rule_id].record(event.source, index)
                if not self._windows[rule.rule_id].is_tripped(hits):
                    # Seen, counted, not yet enough to say anything about.
                    continue

            if winner is None or self._outranks(rule, winner):
                winner, winning_path, winning_hits = rule, match.path, hits

        if winner is None:
            verdict = Verdict(
                event=event,
                decision=self._pack.default_decision,
                reason='No rule matched, so the pack default applied.',
            )
        else:
            verdict = Verdict(
                event=event,
                decision=winner.decision,
                rule=winner,
                path=winning_path,
                reason=self._explain(winner, winning_hits),
            )

        self._counts[verdict.decision] += 1
        if verdict.is_alert:
            self._alerts.append(Alert(verdict, hits=winning_hits))
        return verdict

    def _outranks(self, candidate: Rule, holder: Rule) -> bool:
        """Report whether one matching rule should beat another.

        Higher priority wins. Equal priorities break the tie by order in the
        pack, earlier first, which is the only ordering a rule author can
        actually see and control.

        Parameters
        ----------
        candidate : Rule
            The rule being considered.
        holder : Rule
            The rule currently winning.

        Returns
        -------
        bool
            Whether the candidate should take over.
        """
        if candidate.priority != holder.priority:
            return candidate.priority > holder.priority
        return self._order[candidate.rule_id] < self._order[holder.rule_id]

    @staticmethod
    def _explain(rule: Rule, hits: int) -> str:
        """Write the one-line reason shown with a verdict.

        Parameters
        ----------
        rule : Rule
            The rule that decided.
        hits : int
            How many events were in the window when it fired.

        Returns
        -------
        str
            A sentence a person can read without knowing the rule pack.
        """
        if rule.description:
            base = rule.description
        else:
            base = f'Matched {" ".join(rule.pattern)}.'
        if rule.is_threshold_rule:
            return f'{base} Fired after {hits} matching events from one source.'
        return base

    def classify_events(self, events: Iterable[Event]) -> Iterator[Verdict]:
        """Classify a stream of events, skipping blank lines.

        The events are pulled through a queue rather than a list. Nothing
        here needs random access, and a deque keeps the memory flat no matter
        how long the log is, which is the point of streaming it at all.

        Parameters
        ----------
        events : iterable of Event
            The events to classify, in order.

        Yields
        ------
        Verdict
            One per non-blank line.
        """
        pending: deque[Event] = deque()
        position = 0

        for event in events:
            pending.append(event)
            while pending:
                current = pending.popleft()
                if current.is_blank:
                    continue
                position += 1
                yield self.classify(current, position=position)

    def classify_lines(self, lines: Iterable[str]) -> Iterator[Verdict]:
        """Classify raw log lines.

        Parameters
        ----------
        lines : iterable of str
            Raw lines, with or without trailing newlines. A file object works
            directly, which is what keeps memory flat on a large log.

        Returns
        -------
        iterator of Verdict
            One per non-blank line. This returns the iterator rather than
            yielding, so nothing is read until the caller starts consuming it.
        """
        return self.classify_events(events_from_lines(lines))

    def __str__(self) -> str:
        """Return a one-line summary for the report header."""
        return f'PolicyEngine over {self._pack} using {self._automaton}'

    def __repr__(self) -> str:
        """Return an unambiguous form for debugging."""
        return f'PolicyEngine(pack={self._pack.name!r}, alerts={len(self._alerts)})'

File: web/src/lib/engine.ts
/**
 * The policy engine, ported from policyguard/engine.py.
 *
 * Precedence is the part worth reading twice: when several rules match one
 * line, the highest priority wins, and equal priorities break the tie by
 * order in the pack. Getting that wrong is the easiest way for this port to
 * disagree with the Python while still looking right, which is exactly what
 * the parity check exists to catch.
 */

import { Automaton } from "./automaton.ts";
import { parseEvent, type LogEvent } from "./event.ts";
import { isThresholdRule, type Decision, type Rule, type RulePack } from "./rules.ts";

/** How many recent events a threshold rule keeps when it names no window. */
const DEFAULT_WINDOW_EVENTS = 50;

/** The classification of one log line. Mirrors Verdict.to_dict in Python. */
export interface Verdict {
  line: number;
  raw: string;
  decision: Decision;
  ruleId: string;
  ruleLabel: string;
  severity: string;
  path: number[];
  pathText: string;
  reason: string;
  source: string;
}

/** What classifying a whole log produced. */
export interface Report {
  policy: string;
  counts: Record<Decision, number>;
  verdicts: Verdict[];
  automaton: string;
}

/** Recent hits for one threshold rule, grouped by source. */
class SlidingWindow {
  private readonly hits = new Map<string, number[]>();
  private readonly threshold: number;
  private readonly span: number;

  // Written out rather than using constructor parameter properties, which
  // Node's type stripping refuses, and this runs under `node --test`.
  constructor(threshold: number, span: number) {
    this.threshold = threshold;
    this.span = Math.max(span, threshold);
  }

  /** Record a hit and report how many are inside the window now. */
  record(source: string, position: number): number {
    const queue = this.hits.get(source) ?? [];
    queue.push(position);
    // Only the front can ever be stale, so this stays cheap.
    let start = 0;
    while (start < queue.length && position - (queue[start] as number) >= this.span) start += 1;
    const trimmed = start > 0 ? queue.slice(start) : queue;
    this.hits.set(source, trimmed);
    return trimmed.length;
  }

  /** Report whether a hit count is enough to fire the rule. */
  isTripped(count: number): boolean {
    return count >= this.threshold;
  }
}

/** Write the one-line reason shown with a verdict. */
function explain(rule: Rule, hits: number): string {
  const base = rule.description || `Matched ${rule.pattern.join(" ")}.`;
  if (isThresholdRule(rule)) {
    return `${base} Fired after ${hits} matching events from one source.`;
  }
  return base;
}

/** Render the state path the way --explain prints it. */
function describePath(path: readonly number[], rule: Rule | null): string {
  if (path.length === 0 || rule === null) return "q0 (no rule matched, pack default applied)";
  return `${path.map((state) => `q${state}`).join(" -> ")} -> ACCEPT: ${rule.ruleId}`;
}

/** Classifies log lines against a rule pack. */
export class PolicyEngine {
  private readonly pack: RulePack;
  private readonly automaton: Automaton;
  private readonly order = new Map<string, number>();
  private windows = new Map<string, SlidingWindow>();

  constructor(pack: RulePack) {
    this.pack = pack;
    this.automaton = new Automaton().compile(pack.rules);
    pack.rules.forEach((rule, index) => this.order.set(rule.ruleId, index));
    this.reset();
  }

  /** Forget every sliding window. Call this between files. */
  reset(): void {
    this.windows = new Map(
      this.pack.rules
        .filter(isThresholdRule)
        .map((rule) => [
          rule.ruleId,
          new SlidingWindow(rule.threshold, rule.windowSeconds || DEFAULT_WINDOW_EVENTS),
        ]),
    );
  }

  /** Report whether one matching rule should beat another. */
  private outranks(candidate: Rule, holder: Rule): boolean {
    if (candidate.priority !== holder.priority) return candidate.priority > holder.priority;
    return (this.order.get(candidate.ruleId) ?? 0) < (this.order.get(holder.ruleId) ?? 0);
  }

  /** Classify one event. */
  classify(event: LogEvent, position: number): Verdict {
    let winner: Rule | null = null;
    let winningPath: readonly number[] = [];
    let winningHits = 1;

    for (const match of this.automaton.match(event.tokens)) {
      const rule = match.rule;
      let hits = 1;

      if (isThresholdRule(rule)) {
        const window = this.windows.get(rule.ruleId);
        if (!window) continue;
        hits = window.record(event.source, position);
        if (!window.isTripped(hits)) continue;
      }

      if (winner === null || this.outranks(rule, winner)) {
        winner = rule;
        winningPath = match.path;
        winningHits = hits;
      }
    }

    const decision = winner ? winner.decision : this.pack.defaultDecision;
    return {
      line: event.lineNumber,
      raw: event.raw,
      decision,
      ruleId: winner ? winner.ruleId : "default",
      ruleLabel: winner ? winner.label : "",
      severity: winner && decision === "alert" ? winner.severity : "",
      path: [...winningPath],
      pathText: describePath(winningPath, winner),
      reason: winner ? explain(winner, winningHits) : "No rule matched, so the pack default applied.",
      source: event.source,
    };
  }

  /** Classify raw log text, skipping blank lines. */
  classifyText(text: string): Report {
    this.reset();
    const counts: Record<Decision, number> = { allow: 0, alert: 0, ignore: 0 };
    const verdicts: Verdict[] = [];
    let position = 0;

    // Split the way Python's splitlines does for the shapes a log has, and
    // drop a single trailing empty string so a file ending in a newline does
    // not produce one extra blank event.
    const lines = text.split("\n");
    if (lines.length > 0 && lines[lines.length - 1] === "") lines.pop();

    lines.forEach((raw, index) => {
      const event = parseEvent(raw, index + 1);
      if (event.isBlank) return;
      position += 1;
      const verdict = this.classify(event, position);
      counts[verdict.decision] += 1;
      verdicts.push(verdict);
    });

    return {
      policy: this.pack.name,
      counts,
      verdicts,
      automaton: this.automaton.describe(),
    };
  }
}
